README.md
# Klient Mock

![badge-coverage](.github/badges/coverage.svg)

- [Introduction](#introduction)
- [Setup](#setup)
- [Usage](#usage)
- [Parameters](#parameters)
- [Mock](#mock)

## Introduction

This [Klient](https://github.com/klientjs/core) extension allows you to mock Klient responses. That's means that you can work on an entrypoints even if it's not really ready yet.

## Setup

Install package with your favorite package manager :

```bash
# With NPM
$ npm install @klient/mock

# With YARN
$ yarn add @klient/mock
```

Then import the extension in your code :

```js
import Klient from '@klient/core';

//
// Register extension
//
import '@klient/mock';


//
// Build Klient instance
//
const klient = new Klient('...');


//
// See loaded extension
//
console.log(klient.extensions); // Print ["@klient/mock"]
```

## Usage

```js
import Klient from '@klient/core';

//
// Register extension
//
import '@klient/mock';


//
// Build Klient instance
//
const klient = new Klient(...);


//
// Mock a route - see Mock method doc for more details
//
klient.mock({
  req: { url: '/posts' },         // Request criterias
  res: { status: 200, data: [] }  // Mocked response
});


//
// Consume a mocked route
//
klient.request('/posts').then(response => {
  console.log(response.data); // Print []
});
```

## Parameters

```js
const klient = new Klient({
  delay: 800, // Default response delay to apply to all responses (milliseconds)
  mock: {
    // This parameter is analyzed only on Klient instanciation 
    load: [
      // Mock GET /posts
      {
        delay: 800,  // Simulate response delay (milliseconds)
        req: { url: '/posts' },
        res: { status: 200, data: [] }
      },
    ]
  }
});
```


## Mock

A mock is a fake response build for a request whose matches some criterias.

The mocked response must be an object contening at least the "status" property. It must contains only Axios object properties, that will be "normalized" by mock extension to a final Axios result. The returned status code determines if the request must be rejected or resolved (as Axios does).

To match a mock, the request config must contain all properties/values defined in criterias (it's not strict, request config can contains more properties not defined in criterias). The mock request url can contains dynamic parameters whoses will be treated separately of the rest of config, then they will be injected in mock response callback if match.

```typescript
mock({
  req: AxiosRequestConfig,
  res: AxiosResponse | ((config: AxiosRequestConfig, parameters: object) => AxiosResponse)
}): Klient

// Can be multiple
mock(...Mocks): Klient
```

*Example*

```js
import Klient from '@klient/core';

//
// Register extension
//
import '@klient/mock';


//
// Build Klient instance
//
const klient = new Klient(...);


//
// Mock a route
//
klient.mock({
  req: { url: '/posts' },         // Request criterias
  res: { status: 200, data: [] }  // Mocked response
});


//
// Mock multiple routes
//
klient.mock(
  {
    req: { url: '/posts', method: 'GET' },
    res: { status: 200, data: [] }
  },
  {
    req: { url: '/posts', method: 'POST' },
    res: { status: 400, data: {...} }
  }
);


//
// Mock complexe routes
//
klient.mock({
  delay: 600, // (milliseconds)
  req: { url: '/posts', method: 'POST' },
  // Use a callback to make complex response
  res: (config) => {
    // Simulate API validation errors
    if (!config.data.title) {
      return { status: 400, data: {...} };
    }

    return { status: 200, data: {...} };
  }
});

//
// Mock routes with params
//
klient.mock({
  req: { url: '/posts/{id}', method: 'GET' },
  // Use a callback to make complex response
  res: (config, parameters) => {
    // Simulate API not found
    if (parameters.id !== '1') {
      return {
        status: 404,
        data: {...}
      };
    }

    return { status: 200, data: {...} };
  }
});
```


CHANGELOG.md
# Changelog

## [1.1.2](https://github.com/klientjs/mock/compare/1.1.1...1.1.2) (2023-06-10)


### Features

* apply delay inside request handler function ([f2f0eb2](https://github.com/klientjs/mock/commit/f2f0eb225b26b0bdaff007cc081c1cfc8a5d8602))

## [1.1.1](https://github.com/klientjs/mock/compare/1.1.0...1.1.1) (2023-06-10)


### Features

* override request handler instead of using deprecated resolve and reject method ([5dabac0](https://github.com/klientjs/mock/commit/5dabac0f586a8db278196950d56a7a41e0c44856))

# 1.1.0 (2023-05-14)


### Features

* initialize klient mock extension ([349232f](https://github.com/klientjs/mock/commit/349232fe0c4151d50ec23f5214015e89deb987f6))

tests/mock.test.ts
import Klient, { RequestEvent } from '@klient/core';

import '../src';

import type { Parameters, KlientExtended, Mock } from '../src';

jest.useFakeTimers();
jest.spyOn(global, 'setTimeout');

test('register', () => {
  const klient = new Klient();

  expect(klient.extensions.includes('@klient/mock')).toBeTruthy();
});

test('mock', async () => {
  const klient = new Klient<Parameters>({
    mock: {
      load: [
        {
          req: {
            url: '/auth',
            method: 'POST'
          },
          res: (config) => {
            if (!config.data?.username) {
              return {
                status: 400,
                data: {
                  username: 'this property is required'
                }
              };
            }

            if (!config.data?.password) {
              return {
                status: 400,
                data: {
                  password: 'this property is required'
                }
              };
            }

            return {
              status: 200,
              data: {
                token: '...'
              }
            };
          }
        },
        {
          req: {
            url: '/posts',
            params: {
              test: 1
            }
          },
          res: {
            status: 200,
            data: [
              {
                title: 'test',
                content: 'test'
              }
            ]
          }
        },
        {
          req: {
            url: '/posts/{id}',
            method: 'GET'
          },
          res: (_config, parameters) => {
            return {
              status: 200,
              data: {
                id: parameters.id,
                title: 'test'
              }
            };
          }
        }
      ]
    }
  }) as KlientExtended;

  const mock: Mock = {
    delay: 20,
    req: { url: '/error' },
    res: { status: 500 }
  };

  klient.mock(mock);

  await klient
    .request({
      url: '/auth',
      method: 'POST'
    })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(400);
      expect(e.response.data.username).toBeDefined();
    });

  await klient
    .request({
      url: '/auth',
      method: 'POST',
      data: {
        username: 'test',
        password: 'test'
      }
    })
    .then(({ status, data }) => {
      expect(status).toBe(200);
      expect((data as { token: string }).token).toBe('...');
    })
    .catch((e) => {
      console.log(e);
      throw e;
    });

  await klient
    .request({
      url: '/posts',
      method: 'GET',
      params: {
        test: 1
      }
    })
    .then(({ status }) => {
      expect(status).toBe(200);
    })
    .catch((e) => {
      console.log(e);
      throw e;
    });

  const req = klient.request({ url: '/error' });

  jest.runAllTimers();

  req
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(500);
      expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 20);
    });

  await klient
    .request({ url: '/posts/1', method: 'GET' })
    .then(({ data }) => {
      expect((data as { id: string }).id).toBe('1');
    })
    .catch((e) => {
      console.log(e);
      throw e;
    });

  const cancelRequest = (e: RequestEvent) => {
    e.request.cancel();
  };

  klient.on('request', cancelRequest, -1004);

  await klient
    .request({
      url: '/posts',
      method: 'GET',
      params: {}
    })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(klient.isCancel(e)).toBeTruthy();
    });

  klient.off('request', cancelRequest).on('request', cancelRequest);

  await klient
    .request({ url: '/posts/1', method: 'GET' })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(klient.isCancel(e)).toBeTruthy();
    });
});


src/__tests__/mock.test.ts
import Klient, { RequestEvent } from '@klient/core';

import '..';

import type { Parameters, KlientExtended, Mock } from '..';

jest.useFakeTimers();
jest.spyOn(global, 'setTimeout');

test('register', () => {
  const klient = new Klient();

  expect(klient.extensions.includes('@klient/mock')).toBeTruthy();
});

test('mock', async () => {
  const klient = new Klient<Parameters>({
    mock: {
      load: [
        {
          req: {
            url: '/auth',
            method: 'POST'
          },
          res: (config) => {
            if (!config.data?.username) {
              return {
                status: 400,
                data: {
                  username: 'this property is required'
                }
              };
            }

            if (!config.data?.password) {
              return {
                status: 400,
                data: {
                  password: 'this property is required'
                }
              };
            }

            return {
              status: 200,
              data: {
                token: '...'
              }
            };
          }
        },
        {
          req: {
            url: '/posts',
            params: {
              test: 1
            }
          },
          res: {
            status: 200,
            data: [
              {
                title: 'test',
                content: 'test'
              }
            ]
          }
        },
        {
          req: {
            url: '/posts/{id}',
            method: 'GET'
          },
          res: (_config, parameters) => {
            return {
              status: 200,
              data: {
                id: parameters.id,
                title: 'test'
              }
            };
          }
        }
      ]
    }
  }) as KlientExtended;

  const mock: Mock = {
    delay: 20,
    req: { url: '/error' },
    res: { status: 500 }
  };

  klient.mock(mock);

  await klient
    .request({
      url: '/auth',
      method: 'POST'
    })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(400);
      expect(e.response.data.username).toBeDefined();
    });

  await klient
    .request({
      url: '/auth',
      method: 'POST',
      data: {
        username: 'test',
        password: 'test'
      }
    })
    .then(({ status, data }) => {
      expect(status).toBe(200);
      expect((data as { token: string }).token).toBe('...');
    })
    .catch((e) => {
      console.log(e);
      throw e;
    });

  await klient
    .request({
      url: '/posts',
      method: 'GET',
      params: {
        test: 1
      }
    })
    .then(({ status }) => {
      expect(status).toBe(200);
    })
    .catch((e) => {
      console.log(e);
      throw e;
    });

  const req = klient.request({ url: '/error' });

  jest.runAllTimers();

  req
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(e.response.status).toBe(500);
      expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 20);
    });

  await klient
    .request({ url: '/posts/1', method: 'GET' })
    .then(({ data }) => {
      expect((data as { id: string }).id).toBe('1');
    })
    .catch((e) => {
      console.log(e);
      throw e;
    });

  const cancelRequest = (e: RequestEvent) => {
    e.request.cancel();
  };

  klient.on('request', cancelRequest, -1004);

  await klient
    .request({
      url: '/posts',
      method: 'GET',
      params: {}
    })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(klient.isCancel(e)).toBeTruthy();
    });

  klient.off('request', cancelRequest).on('request', cancelRequest);

  await klient
    .request({ url: '/posts/1', method: 'GET' })
    .then(() => {
      throw new Error('This request must failed');
    })
    .catch((e) => {
      expect(klient.isCancel(e)).toBeTruthy();
    });
});


src/index.ts
import { Extensions } from '@klient/core';

import type Klient from '@klient/core';
import type { Parameters as KlientParameters } from '@klient/core';
import Mocks, { Mock } from './services/mocks';

export { Mock } from './services/mocks';

export interface Parameters extends KlientParameters {
  mock?: {
    delay?: number;
    load?: Mock[];
  };
}

export interface KlientExtended extends Klient<Parameters> {
  mocks: Mocks;
  mock(...collection: Mock[]): KlientExtended;
}

export const extension = {
  name: '@klient/mock',
  initialize: (klient: Klient<Parameters>) => {
    const mocks = new Mocks(klient);

    klient.services.set('mocks', mocks);

    // prettier-ignore
    klient
      .extends('mocks', mocks)
      .extends('mock', (...collection: Mock[]) => {
        mocks.mock(...collection);
        return klient;
      });

    mocks.mock(...((klient.parameters.get('mock.load') || []) as Mock[]));
  }
};

Extensions.push(extension);


src/toolbox/extract-parameters.ts
const extractParameters = (mockUrl: string, requestUrl: string) => {
  const routePaths = mockUrl.split('/').filter((v) => v !== '');
  const requestPaths = requestUrl.split('/').filter((v) => v !== '');
  const parameters: Record<string, unknown> = {};

  // Url parts doesn't match
  if (routePaths.length !== requestPaths.length) {
    return undefined;
  }

  let match = false;

  for (let n = 0, rlen = routePaths.length; n < rlen; n += 1) {
    const routePath = routePaths[n];
    const requestPath = requestPaths[n];

    if (routePath === requestPath) {
      match = n === rlen - 1;
      continue;
    }

    if (routePath.indexOf('{') === 0 && requestPath !== undefined) {
      const paramName = String(routePath).replace('{', '').replace('}', '');
      parameters[paramName] = requestPath;
      match = n === rlen - 1;
    }
  }

  if (match) {
    return parameters;
  }
};

export default extractParameters;


src/services/mocks.ts
import { AxiosError, CanceledError } from 'axios';
import Klient, { RequestEvent, KlientRequestConfig } from '@klient/core';

import type { AxiosResponseHeaders, AxiosResponse } from 'axios';
import statusCodes from '../toolbox/status-codes';
import partialDeepEqual from '../toolbox/partial-deep-equal';
import extractParameters from '../toolbox/extract-parameters';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export interface PartialAxiosResponse<T = any> {
  status: number;
  data?: T;
  headers?: AxiosResponseHeaders;
  config?: KlientRequestConfig;
}

export type ResParameters = Record<string, unknown>;

export type Mock = {
  req: KlientRequestConfig;
  res: PartialAxiosResponse | ((config: KlientRequestConfig, parameters: ResParameters) => PartialAxiosResponse);
  delay?: number;
};

export default class Mocks {
  private mocks: Mock[] = [];

  constructor(private readonly klient: Klient) {
    klient.on('request', this.match.bind(this), -1000);
    klient.on('request', this.handle.bind(this), -Infinity);
  }

  mock(...mocks: Mock[]): this {
    this.mocks.push(...mocks);
    return this;
  }

  private match(e: RequestEvent) {
    const { config } = e;

    for (let i = 0, len = this.mocks.length, parameters: ResParameters = {}; i < len; i += 1) {
      const mock = this.mocks[i];
      const mockReq = { ...mock.req };

      parameters = {};

      if (mockReq.url && config.url) {
        parameters = extractParameters(mockReq.url, config.url) || {};
        // Found parameters means mock has matched request url
        if (parameters && Object.keys(parameters).length > 0) {
          delete mockReq.url;
        }
      }

      if (partialDeepEqual(mockReq, config)) {
        e.context.mock = mock;
        e.context.parameters = parameters;

        break;
      }
    }
  }

  private handle(e: RequestEvent) {
    const { context, request } = e;
    const { mock } = context as { mock?: Mock };

    if (!mock) {
      return;
    }

    const delay = mock.delay !== undefined ? mock.delay : this.klient.parameters.get('mock.delay');

    request.handler = () =>
      new Promise((resolve, reject) => {
        const callback = () => {
          try {
            resolve(Mocks.renderResponse(e));
          } catch (err) {
            reject(err);
          }
        };

        if (typeof delay === 'number') {
          setTimeout(callback, delay);
        } else {
          callback();
        }
      });
  }

  private static renderResponse(e: RequestEvent) {
    const { config, context } = e;
    const { mock, parameters } = context as { mock: Mock; parameters: unknown };

    // Cancel
    if ((config.signal as AbortSignal).aborted) {
      const err = new CanceledError();

      err.message = AxiosError.ERR_CANCELED;
      err.code = AxiosError.ERR_CANCELED;
      err.config = e.config;

      throw err;
    }

    const result = typeof mock.res === 'function' ? mock.res(config, parameters as ResParameters) : mock.res;
    const { status } = result;

    // Success
    if (status >= 200 && status < 400) {
      return result as AxiosResponse;
    }

    // Error
    const err = new AxiosError();

    err.message = statusCodes.getStatusText(status);
    err.code = AxiosError.ERR_BAD_RESPONSE;
    err.config = e.config;
    err.response = result as AxiosResponse;

    throw err;
  }
}
